isolate/mtn.go

```diff
@@ -0,0 +1,294 @@
+package isolate
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+	"sync"
+)
+
+type RawAlloc struct {
+	Id string `json:"id"`
+	Porto RawPorto `json:"porto"`
+	Network string `json:"network"`
+}
+
+type RawPorto struct {
+	Hostname string `json:"hostname"`
+	Ip string `json:"ip"`
+	Net string `json:"net"`
+	Id string `json:",omitempty"`
+}
+
+type AllocError struct {
+        Type string `json:"type,omitempty"`
+        Message []string `json:"message"`
+        Cause ErrorCause `json:"cause"`
+}
+
+type ErrorCause struct {
+        Type string `json:"type,omitempty"`
+        Message []string `json:"message"`
+        Cause string `json:"cause,omitempty"`
+}
+
+type RawAllocs []RawAlloc
+
+type AllocAnswer string
+
+type MtnCfg struct {
+	Enable bool
+	Allocbuffer int
+	Url string
+	Ident string
+	SchedLabel string
+	Headers map[string]string
+}
+
+type MtnState struct {
+	sync.Mutex
+	Cfg MtnCfg
+	Pool MtnPool
+}
+
+type MtnPool map[string]*IdState
+
+type IdState struct {
+	sync.Mutex
+	Reserved int
+	Allocations map[string]Allocation
+}
+
+type Allocation struct {
+	Net string
+	Hostname string
+	Ip string
+	Id string
+	Used bool
+}
+
+type PostAllocreq struct {
+	Network string `json:"network"`
+	Host string `json:"host"`
+	Scheduler string `json:"scheduler"`
+}
+
+func (c *MtnState) CfgInit(cfg *Config) bool {
+	c.Cfg.Enable = cfg.Mtn.Enable
+	if !c.Cfg.Enable {
+		return true
+	}
+	if cfg.Mtn.Allocbuffer < 1 {
+		c.Cfg.Allocbuffer = 3
+	} else {
+		c.Cfg.Allocbuffer = cfg.Mtn.Allocbuffer
+	}
+	if len(cfg.Mtn.Label) == 0 {
+		fqdn, err := os.Hostname()
+		if err != nil {
+			fmt.Println("%s ERROR: Cant get hostname inside CfgInit() by calling os.Hostname(), returned: %s", time.Now().UTC().Format(time.RFC3339), err)
+			return false
+		}
+		c.Cfg.SchedLabel = fqdn
+	} else {
+		c.Cfg.SchedLabel = cfg.Mtn.Label
+	}
+	if len(cfg.Mtn.Ident) == 0 {
+		c.Cfg.Ident = c.Cfg.SchedLabel
+	} else {
+		c.Cfg.Ident = cfg.Mtn.Ident
+	}
+	c.Cfg.Url = cfg.Mtn.Url
+	c.Cfg.Headers = cfg.Mtn.Headers
+	return true
+}
+
+func (c *MtnState) PoolInit() bool {
+	if !c.Cfg.Enable {
+		return true
+	}
+	c.Pool = make(map[string]*IdState)
+	allAllocs, err := c.GetAllocations()
+	if err != nil {
+		// TODO: maybe need use main logger at this step
+		fmt.Println("%s ERROR: Cant init pool inside PoolInit(), err: %s", time.Now().UTC().Format(time.RFC3339), err)
+		return false
+	}
+	for netId, allocs := range allAllocs {
+		cState := *new(IdState)
```

Comment:
? why deref of pointer? just var cState IdState

Reply:
fixed, tnx